netty/src/main/java/io/grpc/netty/ProtocolNegotiators.java

```diff
@@ -407,31 +407,79 @@ public HostPort(String host, int port) {
+    return new PlaintextUpgradeProtocolNegotiator();
   }
 
-  static final class PlaintextUpgradeNegotiator implements ProtocolNegotiator {
+  static final class PlaintextUpgradeProtocolNegotiator implements ProtocolNegotiator {
 
     @Override
-    public ChannelHandler newHandler(GrpcHttp2ConnectionHandler handler) {
-      // Register the plaintext upgrader
-      Http2ClientUpgradeCodec upgradeCodec = new Http2ClientUpgradeCodec(handler);
-      HttpClientCodec httpClientCodec = new HttpClientCodec();
-      final HttpClientUpgradeHandler upgrader =
-          new HttpClientUpgradeHandler(httpClientCodec, upgradeCodec, 1000);
-      return new BufferingHttp2UpgradeHandler(httpClientCodec, upgrader, handler,
-          handler.getAuthority());
+    public AsciiString scheme() {
+      return Utils.HTTP;
     }
 
     @Override
-    public AsciiString scheme() {
-      return Utils.HTTP;
+    public ChannelHandler newHandler(GrpcHttp2ConnectionHandler grpcHandler) {
+      ChannelHandler upgradeHandler =
+          new Http2UpgradeAndGrpcHandler(grpcHandler.getAuthority(), grpcHandler);
+      ChannelHandler wuah = new WaitUntilActiveHandler(upgradeHandler);
+      return wuah;
     }
 
     @Override
     public void close() {}
   }
 
+  /**
+   * Acts as a combination of Http2Upgrade and {@link GrpcNegotiationHandler}.  Unfortunately,
+   * this negotiator doesn't follow the pattern of "just one handler doing negotiation at a time."
+   * This is due to the tight coupling between the upgrade handler and the HTTP/2 handler.
+   */
+  static final class Http2UpgradeAndGrpcHandler extends ChannelInboundHandlerAdapter {
+
+    private final String authority;
+    private final GrpcHttp2ConnectionHandler next;
+
+    private ProtocolNegotiationEvent pne = ProtocolNegotiationEvent.DEFAULT;
+
+    Http2UpgradeAndGrpcHandler(String authority, GrpcHttp2ConnectionHandler next) {
+      this.authority = checkNotNull(authority, "authority");
+      this.next = checkNotNull(next, "next");
+    }
+
+    @Override
+    public void handlerAdded(ChannelHandlerContext ctx) throws Exception {
+      HttpClientCodec httpClientCodec = new HttpClientCodec();
+      ctx.pipeline().addBefore(ctx.name(), null, httpClientCodec);
+
+      Http2ClientUpgradeCodec upgradeCodec = new Http2ClientUpgradeCodec(next);
+      HttpClientUpgradeHandler upgrader =
+          new HttpClientUpgradeHandler(httpClientCodec, upgradeCodec, /*maxContentLength=*/ 1000);
+      ctx.pipeline().addBefore(ctx.name(), null, upgrader);
+
+      // Trigger the HTTP/1.1 plaintext upgrade protocol by issuing an HTTP request
+      // which causes the upgrade headers to be added
+      DefaultHttpRequest upgradeTrigger =
+          new DefaultHttpRequest(HttpVersion.HTTP_1_1, HttpMethod.GET, "/");
+      upgradeTrigger.headers().add(HttpHeaderNames.HOST, authority);
+      ctx.writeAndFlush(upgradeTrigger).addListener(FireExceptionListener.INSTANCE);
```

Comment:
ChannelFutureListener.FIRE_EXCEPTION_ON_FAILURE

Reply:
TIL.  

I remembered the CLOSE one, but didn't realize the exception one.